### PR TITLE
Make layer private

### PR DESCRIPTION
**Pull request by mariaschuld**
*Note: This PR is forked from PR #409 which should be merged first*.

**Context:** Part of the templates refactor. Since ``Layer()`` functions are contained in ``Layers()`` by setting the number of layers to 1, we can discard them.

**Description of the Change:** Turns all ``Layer()`` templates into private functions.

**Benefits:** Less clutter, clearer signatures.

**Possible Drawbacks:** Feature breaking.

**Related GitHub Issues:** -

### CODE REVIEW

**Comment by mariaschuld** [comment]
This relatively small PR would be very convenient to merge before later ones. It also contains doc improvements!

File: tests/test_init.py [hunk 1]
```diff
@@ -17,19 +17,13 @@
 # pylint: disable=protected-access,cell-var-from-loop
 import pytest
```
Comment (mariaschuld) [inline tests/test_init.py]: The following removes all `layer` - related initializations and their tests

File: tests/test_templates_subroutines.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+        @qml.qnode(dev)
+        def circuit(varphi, mesh):
+            Interferometer(theta=None, phi=None, varphi=varphi, mesh=mesh, wires=0)
+            return qml.expval(qml.NumberOperator(0))
+
+        with pytest.raises(QuantumFunctionError) as excinfo:
+            circuit(varphi, 'rectangular')
+        assert excinfo.value.args[0] == 'The mesh parameter influences the circuit architecture ' \
+                                        'and can not be passed as a QNode parameter.'
```
Comment (josh146) [inline tests/test_templates_subroutines.py]: You can use `pytest.raises(QuantumFunctionError, match="can not be passed"):`, which avoids the additional line, and allows you to use regex to match the string instead of matching the exact string.

File: tests/test_templates_subroutines.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+        res = circuit(theta, phi, varphi)
+        expected = np.array([0.96852694, 0.23878521, 0.82310606, 0.16547786])
+        assert np.allclose(res, expected, atol=tol)
+
+        res = qml.jacobian(circuit, 0)(theta, phi, varphi)
+        expected = np.array([[-6.18547248e-03, -3.20488426e-04, -4.20274087e-02, -6.21819638e-02, 9.68526932e-01, 9.68526932e-01],
+                             [ 3.55439246e-04,  3.89820238e-02, -3.35281306e-03,  7.93009027e-04, 8.30347888e-02,-3.45150707e-01],
+                             [ 5.44893380e-03,  9.30878007e-03, -5.33374094e-01,  6.13889548e-02, -1.16931385e-01, 3.45150707e-01],
+                             [ 3.81099442e-04, -4.79703154e-02,  5.78754316e-01,  1.65477867e-01, 3.38965967e-02, 1.65477867e-01]])
```
Comment (josh146) [inline tests/test_templates_subroutines.py]: How do we know this is correct/where is this from?

Ideally, we should have an analytic expression here, _or_ commands/comments showing how to reproduce this result

File: tests/test_templates_subroutines.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+"""
+# pylint: disable=protected-access,cell-var-from-loop
+import pytest
+
+import logging as log
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.qnode import QuantumFunctionError
+from pennylane.templates.layers import (Interferometer)
+log.getLogger('defaults')
```
Comment (josh146) [inline tests/test_templates_subroutines.py]: Is the logger used in this test?

File: tests/test_templates.py [hunk 5]
```diff
@@ -12,106 +12,351 @@
...
+            for a in argnm:
+                inpts[a].grad.numpy()
 
-        p = parfun(n_wires=n_subsystems)
+        # Check gradients in tf interface
+        if intrfc == 'tf':
+            grad_inpts = [inpts[a] for a in argnm]
+            with tf.GradientTape() as tape:
+                loss = circuit(*inpts)
+                tape.gradient(loss, grad_inpts)
```
Comment (josh146) [inline tests/test_templates.py]: Should probably assert that these gradient values are not `None`

File: tests/test_templates.py [hunk 6]
```diff
@@ -12,106 +12,351 @@
...
+            res.backward()
+            for a in argnm:
+                inpts[a].grad.numpy()
+
+        # Check gradients in tf interface
+        if intrfc == 'tf':
+            grad_inpts = [inpts[a] for a in argnm]
+            with tf.GradientTape() as tape:
+                loss = circuit(*inpts)
+                tape.gradient(loss, grad_inpts)
```
Comment (josh146) [inline tests/test_templates.py]: great that this works though!

File: tests/test_templates_subroutines.py (hunk 3 above)
Comment (mariaschuld) [reply tests/test_templates_subroutines.py]: Will make new issue.

File: .github/CHANGELOG.md [hunk 7]
```diff
@@ -57,6 +69,10 @@
 
 This release contains contributions from (in alphabetical order):
 
+<<<<<<< HEAD
+
+=======
+>>>>>>> master
```
Comment (co9olguy) [inline .github/CHANGELOG.md]: Looks like some git stuff got merged in??

Comment (josh146) [reply .github/CHANGELOG.md]: @mariaschuld, could you undo this in the next template PR that is to be merged?

Comment (mariaschuld) [reply .github/CHANGELOG.md]: I did, sorry for that! I also checked now if there are any <<<< or HEAD statements in path and it's clean.
